### ISSUE TITLE
Missing 'using' for lifted compiler-generated helper

### ISSUE THREAD

**Issue by tamlin-mike** [icsharpcode/ILSpy#2609]
Tested with 65792dfa5. Documenting it in case it's still present in HEAD.

If a type has a compiler-generated helper that uses e.g. an extension method from a namespace 'X2' otherwise not needed by the "main" type being decompiled, `using X2;` (could? must?) still be included in the main type's `using`'s.

This can be observed if decompiling the compiler-generated class and looking at its `using`'s. If they include something not part of the main type being decompiled, the main type might (will?) need to include such missing namespace(s).

From the top of my head it went something like this:

When looking at the decompilation of the "main" type
```
using X1;
class C {
  SomeType Method() { return complex_expression1; }
}
```
where a part of `complex_expression` resulted in a compiler-generated class using an extension method from namespace X2 like
```
using X1;
using X2;
[CompilerGenerated]
class CompilerGenerated {
  SomeType Method(...) { return expression_using_type_from_X2; }
}
```

It gave me quite a runaround before I figured out why on earth it faild to re-compile! :-)

**Comment by siegfriedpammer** [comment]
Can you provide the assembly you tried to decompile and re-compile? I am not sure, whether I understand you correctly. You are trying to export an assembly as project and then try to re-compile it? If that's the case, why would you have a compiler-generated nested class in the first place? Is there some other problem, which causes this? I wouldn't expect the compiler-generated class to compile in any case.

The problem you found is that there is an extension method used in the compiler-generated code, which is not imported? If a method definition is not found, the decompiler should not emit the call using extension method syntax, as it will only do so, if it can guarantee that the method is accessible and actually usable without conflicts...

Without a concrete reproducer this will be impossible to investigate. Sorry!

**Comment by tamlin-mike** [comment]
Let me try to shed some light on this. :-)

The compiler-generated class is in the original assembly. It is not present in the decompiled C# source. The specifics were an `await` expression, where a UDT was using an extension method from a type inside namespace X.

When using ILSpy to look at the decompiled C# of the generated FSM, `using X;` was properly included in the decompiled output's using's. The original C# source code that the compiler had moved into the FSM's `MoveNext` used an extension from namespace X.

When using ILSpy to look at the decompiled C# of the "actual" class `C` (the one using the FSM), `using X;` was improperly _not_ included, most likely since `C` did not otherwise reference anything from `X`.

Basically, when ILSpy turns a compiler-generated FSM back into its source form, but there are types referenced from the generated FSM that are otherwise not directly referenced from the class containing the FSM, those using' s are in error not included in the final C# output.

Since I don't recall exactly where that offending code was, let me try to create a pseudo-code example to amend the explanation.

Original source:
```
namespace Z {
  using X;
  class C {
    Task foo() {
      SomeType blah = ...; // unimportant, except to know it's not from namespace X
      return await blah.Baz(); //
    }
  }
}
namespace X {
  public static class Extensions {
    public static Task Baz(this SomeType x) ... // impl unimportant
  }
}
```

In that code, the only mention of namespace `X` is the single usage of the extension method `Baz`.

When using iLSpy interactively to look at the FSM's `MoveNext`, it properlu include `using X;`.

When emitting decompiled source, where the FSM is not included since it has fully been turned into the source expression, `using X; is missing.

Makes sense now?

**Comment by siegfriedpammer** [comment]
![image](https://user-images.githubusercontent.com/344204/150030104-daada868-9020-496f-8011-fb6e3948837a.png)

Cannot reproduce. :(

**Comment by tamlin-mike** [comment]
This should teach me to immediately stop work and properly document a problem...

It's a rather code base where this problem (as I recall it) occurred, and now I simply can't find where it was. Heck, right now I'm starting to think it might have been a missing assembly import (since many assemblies implement stuff in the same namespaces, most likely including the namespace the extension was implemented in). That was a problem, that the original assemblies had imports that were not present in the ILSpy generated project files, and right now I'm starting to wonder if _that_ is the problem I remember.

So maybe, just maybe, this only happens if namespace `X` have classes implemented in a number of different assemblies, and the extension method happens to be in an assembly no "main" code is directly using.

Whatever. I feel I have wasted enogh of your time on this, but if anything comes to mind regarding dropped _assembly_, not namespace, maybe this issue could provide a clue.

Again, I'm really sorry for the confusion and time wasted.
